soroban-env-host/src/native_contract/token/allowance.rs

```diff
@@ -21,17 +28,91 @@ pub fn write_allowance(
     from: Address,
     spender: Address,
     amount: i128,
+    expiration: u32,
 ) -> Result<(), HostError> {
     let key = DataKey::Allowance(AllowanceDataKey { from, spender });
-    e.put_contract_data(
-        key.try_into_val(e)?,
-        amount.try_into_val(e)?,
-        StorageType::Persistent,
-        ().into(),
-    )?;
+
+    // validates the expiration and then returns the ledger seq
+    // The expiration can be less than ledger seq if clearing an allowance
+    let ledger_seq = e.with_ledger_info(|li| {
+        if expiration > li.sequence_number.saturating_add(li.max_entry_expiration) {
+            Err(err!(
+                e,
+                ContractError::AllowanceError,
+                "expiration is greater than max: {} > {}",
+                expiration,
+                li.max_entry_expiration
+            ))
+        } else if amount > 0 && expiration <= li.sequence_number {
+            Err(err!(
+                e,
+                ContractError::AllowanceError,
+                "expiration must be greater than ledger sequence: {} <= {}",
+                expiration,
+                li.sequence_number
+            ))
+        } else {
+            Ok(li.sequence_number)
+        }
+    })?;
+
+    if let Ok(allowance) = e.get_contract_data(key.try_into_val(e)?, StorageType::Temporary) {
+        let mut updated_allowance: AllowanceValue = allowance.try_into_val(e)?;
+
+        updated_allowance.amount = amount;
+        let old_expiration = updated_allowance.expiration_ledger;
+        updated_allowance.expiration_ledger = expiration;
+
+        e.put_contract_data(
+            key.try_into_val(e)?,
+            updated_allowance.try_into_val(e)?,
+            StorageType::Temporary,
+            ().into(),
+        )?;
+
+        if old_expiration < expiration && amount > 0 {
+            e.bump_contract_data(
+                key.try_into_val(e)?,
+                StorageType::Temporary,
+                (expiration - ledger_seq).into(),
```

Comment:
Shall we provide an alternative bump function that uses the exact ledger? This seems error-prone to me (I can't really tell if this is correct or there should be +1). The use case of bumping to exact ledger seems useful for most of the temp entry kinds (as they usually would be associated with absolute ledger boundaries).

Reply:
I think It sounds like the issue is if it's exclusive or not, and if we use the current ledger or lcl (in this case it's current ledger because that's what we pass into `LedgerInfo`.  Using the exact ledger would get rid of the second issue, but you still have the first. I think good documentation should be sufficient (and we can add the other bump function if we need to later).

We can also update the sdk and add another function for the exact ledger (doesn't help here though).

Reply:
Yeah, I'm fine with this as SDK function. I'm sure that it will be a common issue for temp entries otherwise.